### PR TITLE
Implement community-detail-token-gating component & use it in permissions sheet

### LINKED ISSUE #18995: [Communities] Implement Community Detail Token Gating component
Follow up #18638 

[Figma design system](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=35933-64264&mode=design&t=tOndjZ5fVHct6hh1-0)

### Description
While implementing the permissions drawer, it was discovered that our design system lacks a specific component essential for this feature. As a result, the feature was implemented without the integration of this missing component.

### Design

image

### Acceptance Criteria

- Follow Figma
- Pass correct props
- Keep interactions
- Follow [guidelines](https://github.com/status-im/status-mobile/pull/18926/files)
### Notes

Probably the current code can be reused and adapted.

### PR DESCRIPTION
**Pull request by ajayesivan**
fixes #18995 

### Summary
- Implement community/community-detail-token-gating component
   Figma: https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=35933-64264&mode=design&t=R8Hybvk5UZNj9acx-4
- Use the community-detail-token-gating component in the permissions sheet
  Figma: https://www.figma.com/file/h9wo4GipgZURbqqr1vShFN/Communities-for-Mobile?type=design&node-id=15348-171382&mode=design&t=15thOmnbmMzlcFWV-4

#### Platforms

- Android
- iOS

#### Areas that maybe impacted

- Quo Preview -> community -> community-detail-token-gating
- Token gated community permissions sheet

### Steps to test
- Quo Preview -> community -> community-detail-token-gating
----
- Open a token gated community
- Press request to join (user must be eligible to join)
- In the request to join drawer press the permissions button in the top right corner

### Before and after screenshots comparison

| Figma |  Before    | After |
| --- | --- | --- |
| image | Simulator Screenshot - iPhone 13 - 2024-04-05 at 04 09 53 | Simulator Screenshot - iPhone 13 - 2024-04-05 at 04 09 19 |

status: ready

### CODE REVIEW

File: src/quo/components/community/community_detail_token_gating/view.cljs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            [quo.components.tags.collectible-tag.view :as collectible-tag]
+            [quo.components.tags.token-tag.view :as token-tag]
+            [react-native.core :as rn]
+            [utils.i18n :as i18n]))
+
+(defn- token-view
+  [{:keys [collectible? img-src amount sufficient?] :as token}]
+  (if collectible?
+    [collectible-tag/view
+     {:collectible-name    (:symbol token)
```
Comment (ibrkhalil) [inline src/quo/components/community/community_detail_token_gating/view.cljs]: `symbol` can be destructured in line 12

File: src/quo/components/community/community_detail_token_gating/view.cljs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+       :style {:margin-bottom 8}}
+      (i18n/label (if (:satisfied? role) :t/you-hodl :t/you-must-hold))]
+
+     (map-indexed (fn [index tokens]
+                    ^{:key (str (:role role) "-tokens-" index)}
+                    [tokens-row
+                     {:tokens   tokens
+                      :first?   (= index 0)
+                      :divider? (= index (dec (count (:tokens role))))}])
+                  (:tokens role))]))
```
Comment (ibrkhalil) [inline src/quo/components/community/community_detail_token_gating/view.cljs]: A lot can be destructured here, It'll make the function more readable.

File: src/quo/components/community/community_detail_token_gating/view.cljs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  [:<>
+   [rn/view
+    {:style (style/token-row first?)}
+    (map-indexed (fn [token-index token]
+                   ^{:key (str "token-" token-index)}
+                   [token-view token])
+                 tokens)]
+   (when-not divider?
+     [divider-label/view
+      {:container-style {:padding-left 0}}
```
Comment (ibrkhalil) [inline src/quo/components/community/community_detail_token_gating/view.cljs]: Do we need this zero value?

**Review by ibrkhalil — APPROVED** [review]
lgtm, non blocking comments

File: src/quo/components/community/community_detail_token_gating/view.cljs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+     [text/text
+      {:size  :paragraph-2
+       :style {:margin-bottom 8}}
+      (i18n/label (if (:satisfied? role) :t/you-hodl :t/you-must-hold))]
+
+     (map-indexed (fn [index tokens]
+                    ^{:key (str (:role role) "-tokens-" index)}
+                    [tokens-row
+                     {:tokens   tokens
+                      :first?   (= index 0)
```
Comment (ibrkhalil) [inline src/quo/components/community/community_detail_token_gating/view.cljs]: Can be `(zero? index)`

File: src/quo/components/community/community_detail_token_gating/view.cljs (hunk 1 above)
Comment (ajayesivan) [reply src/quo/components/community/community_detail_token_gating/view.cljs]: Destructing `symbol` directly will result in `Shadowed var: cljs.core/symbol` lint error.

File: src/quo/components/community/community_detail_token_gating/view.cljs (hunk 3 above)
Comment (ajayesivan) [reply src/quo/components/community/community_detail_token_gating/view.cljs]: The `divider-label` component has a default left padding of `20px`, but according to the design in this situation, we don't need any padding. That's why I'm overriding it by passing it in the container style.

File: src/quo/components/community/community_detail_token_gating/view.cljs (hunk 1 above)
Comment (ibrkhalil) [reply src/quo/components/community/community_detail_token_gating/view.cljs]: We can instead destructure it like this
```clj
[{token-symbol :symbol} :as token}]
```

File: src/quo/components/community/community_detail_token_gating/view.cljs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      :token-img-src img-src
+      :options       (when sufficient? :hold)}]))
+
+(defn- tokens-row
+  [{:keys [tokens divider? first?]}]
+  [:<>
+   [rn/view
+    {:style (style/token-row first?)}
+    (map-indexed (fn [token-index token]
+                   ^{:key (str "token-" token-index)}
```
Comment (BalogunofAfrica) [inline src/quo/components/community/community_detail_token_gating/view.cljs:35]: Index might not be the best way to key a list. Perhaps a more unique value from the `token` map should be used

File: src/quo/components/community/community_detail_token_gating/view.cljs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      (if satisfied?
+        (i18n/label :t/you-eligible-to-join-as {:role role-text})
+        (i18n/label :t/you-not-eligible-to-join))]
+     [text/text
+      {:size  :paragraph-2
+       :style {:margin-bottom 8}}
+      (i18n/label (if satisfied? :t/you-hodl :t/you-must-hold))]
+
+     (map-indexed (fn [index tokens-item]
+                    ^{:key (str role "-tokens-" index)}
```
Comment (BalogunofAfrica) [inline src/quo/components/community/community_detail_token_gating/view.cljs:64]: Same here with the index

File: src/quo/components/community/community_detail_token_gating/view.cljs (hunk 3 above)
Comment (J-Son89) [reply src/quo/components/community/community_detail_token_gating/view.cljs]: It's probably better to do the reverse?
i.e default is 0 and the other uses have 20 set

Comment (BalogunofAfrica) [reply src/quo/components/community/community_detail_token_gating/view.cljs]: > It's probably better to do the reverse? i.e default is 0 and the other uses have 20 set

That might involve changing it in a number of places I think

Comment (ajayesivan) [reply src/quo/components/community/community_detail_token_gating/view.cljs]: The divider component design has left padding. For this screen only it is being removed. I think it is better to keep the component aligned with the design system.

Design: https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=771-26244&mode=dev

File: src/quo/components/community/community_detail_token_gating/view.cljs (hunk 5 above)
Comment (ajayesivan) [reply src/quo/components/community/community_detail_token_gating/view.cljs:35]: The db doesn't have any unique value for tokens here. Also, the chances of the items changing after rendering are very rare, so I thought it is okay to use the index here.

File: src/quo/components/community/community_detail_token_gating/view.cljs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            [quo.components.dividers.divider-label.view :as divider-label]
+            [quo.components.markdown.text :as text]
+            [quo.components.tags.collectible-tag.view :as collectible-tag]
+            [quo.components.tags.token-tag.view :as token-tag]
+            [react-native.core :as rn]
+            [utils.i18n :as i18n]))
+
+(defn- token-view
+  [{:keys [collectible? img-src amount sufficient?] :as token}]
+  (let [token-symbol (:symbol token)]
```
Comment (ajayesivan) [inline src/quo/components/community/community_detail_token_gating/view.cljs:14]: @ibrkhalil this is what I ended up doing. Let me know if there is a better way to handle this.

**Comment by pavloburykh** [comment]
@ajayesivan thanks for the PR!

There is an issue which also reproducible in develop branch. I am wondering if you can take a look in this PR. If not - I can log separately. WDYT @ajayesivan? 

### ISSUE 1 Not eligible roles' permissions are displayed within community detail bottom sheet

**Steps:** 

1. Create community with permissions Membership and Admin/Token Master permissions. 
2. Open community by user who meets Admin/Token Master permissions requirements but does not meet Membership permission. i.e. user is eligible to join as Admin/Token master but is not eligible to join as Member. 
3. See if not eligible Member role is displayed within community detail bottom sheet

**Actual result:** not eligible Member role is displayed within community detail bottom sheet

photo_2024-04-12 17 21 33

**Expected result:** only eligible roles should be listed within community detail bottom sheet

**Comment by pavloburykh** [comment]
@ajayesivan small UPDATE on [ISSUE 1 ](https://github.com/status-im/status-mobile/pull/19532#issuecomment-2051886742): bug happens in case if user does not have tokens required for Membership permission but holds tokens for same higher role, i.e Admin or Token master. I have updated ISSUE 1 description.

**Comment by pavloburykh** [comment]
@ajayesivan another issue which exists both is PR and develop but affects community-detail-token-gating. Please, tell me if it makes sense to take a look in this PR or should I log separately. 

### ISSUE 2 Community-detail-token-gating screen is broken in case of long scrollable details list

Steps: 

1. Generate a long Permissions list
2. Open community detail token gating bottom sheet
3. Observe the result

**Actual result:** Community-detail-token-gating screen is broken


https://github.com/status-im/status-mobile/assets/97245802/2b66a3fb-e6a6-40e9-8ac8-1970b112412e

**Comment by pavloburykh** [comment]
@ajayesivan there are no more issues besides ISSUE 1 and ISSUE 2 from my side. So if you consider that we should fix them separately - then we can move PR forward. Does it require design review BTW, what do you think @ajayesivan?

**Comment by ajayesivan** [comment]
@pavloburykh Regarding issue 1, from my understanding we should always show membership permissions even if the users don't meet the requirement. The only exception is when the permission is explicitly marked as hidden - [see this issue](https://github.com/status-im/status-mobile/issues/19606).

**Comment by ajayesivan** [comment]
I will fix issue 2 in this PR. Will ping you once it's fixed.

**Comment by ajayesivan** [comment]
One goal of this PR is to make the permission sheet perfectly aligned with the Figma design. So I think we should do a design review.

**Comment by pavloburykh** [comment]
> I will fix issue 2 in this PR. Will ping you once it's fixed.

@ajayesivan thank you!

> @pavloburykh Regarding issue 1, from my understanding we should always show membership permissions even if the users don't meet the requirement. 

My understanding was different. I will catch up with design team to make sure which behaviour is correct. 
It seems weird that we are showing 'Become a member' permissions which user does not meet while he meets higher role permissions (like Admin or Token Master). At the same time, in case if user meet 'Become a member permission' we do not show him 'Admin' permission if he does not meets requirements. So something is definitely wrong here. 

The [issue](https://github.com/status-im/status-mobile/issues/19606) you are reffering to is slightly different and relates to a case when user meets at least some of multiple permissions which are set for specific role. 

Also it looks like that in this PR there is some kind of regression: as you can see on screenshot of [ISSUE 1](https://github.com/status-im/status-mobile/pull/19532#issuecomment-2051886742) 'You are not eligible to join' text misses the role, i.e. 'You are not eligible to join **as Member**'. But let me first ask designers if this behaviour is correct overall.

**Comment by pavloburykh** [comment]
> @pavloburykh Regarding issue 1, from my understanding we should always show membership permissions even if the users don't meet the requirement. 

Hey @ajayesivan! I have contacted the design team and they confirmed that behaviour is correct. 

Could you please check 'You are not eligible to join' text misses the role, should be 'You are not eligible to join **as Member**'? I am not reproducing this issue in develop though I am wondering why would it be PR's regression. 

photo_2024-04-12 17 21 33

**Comment by ajayesivan** [comment]
@pavloburykh I have fixed all the other issues. Could you please take another look? Thank you!

**Comment by pavloburykh** [comment]
@ajayesivan thanks a lot for PR and fixes. Ready for design review cc @Francesca-G

**Review by Francesca-G — COMMENTED** [review]
[Here](https://www.figma.com/file/e9woQowfaIOCBrXS1jgkhZ/Review-Components-%2F-Mobile?type=design&node-id=190%3A4699&mode=design&t=ucPBWVSBWxJglcKX-1) is the design review :)

**Comment by ajayesivan** [comment]
@Francesca-G thank you for the review! I have fixed the issues. Could you please have another look.

File: src/quo/components/community/community_detail_token_gating/style.cljs [hunk 8]
```diff
@@ -12,3 +12,9 @@
    :column-gap     8
    :flex-wrap      :wrap
    :margin-bottom  16})
+
+(def divider
+  {:padding-left 0
+   :height       28
+   :padding-top  0
+   :align-items  :flex-start})
```
Comment (ibrkhalil) [inline src/quo/components/community/community_detail_token_gating/style.cljs:20]: Do need zero values?

Comment (ajayesivan) [reply src/quo/components/community/community_detail_token_gating/style.cljs:20]: Yes, to override the default values of the component.

**Review by Francesca-G — APPROVED** [review]
Looks good now 👍

**Pull request merged**